File: algorithms/[368]最大整除子集/solution.ts
// topics = ["动态规划"]

function largestDivisibleSubset(nums: number[]): number[] {
  const n = nums.length;
  nums.sort((a, b) => a - b);

  // 第 1 步：动态规划找出最大子集的个数、最大子集中的最大整数
  const dp = new Array<number>(n).fill(1);
  let maxSize = 1;
  let maxVal = dp[0];
  for (let i = 1; i < n; i += 1) {
    for (let j = 0; j < i; j += 1) {
      // 「没有重复元素」很重要
      if (nums[i] % nums[j] == 0) {
        dp[i] = Math.max(dp[i], dp[j] + 1);
      }
    }

    if (dp[i] > maxSize) {
      maxSize = dp[i];
      maxVal = nums[i];
    }
  }

  // 第 2 步：倒推获得最大子集
  const res: number[] = [];
  if (maxSize === 1) {
    res.push(nums[0]);
    return res;
  }

  for (let i = n - 1; i >= 0 && maxSize > 0; i -= 1) {
    if (dp[i] == maxSize && maxVal % nums[i] == 0) {
      maxVal = nums[i];
      maxSize -= 1;
      res.push(nums[i]);
    }
  }

  return res;
}

File: algorithms/[368]最大整除子集/solution.py
# topics = ["动态规划"]

from typing import List


class Solution:
    def largestDivisibleSubset(self, nums: List[int]) -> List[int]:
        """
        近似题 300.最长递增子序列
        """
        n = len(nums)
        nums.sort()

        # dp[i] 表示在 [0, i] 区间，必须包含 nums[i] 的最大整数子集个数
        dp = [1] * n

        for i in range(n):
            for j in range(i):
                if nums[i] % nums[j] == 0:
                    dp[i] = max(dp[i], dp[j] + 1)

        # 最大子集的长度
        maxc = max(dp)
        # 最大子集中的最大值
        maxv = nums[dp.index(maxc)]
        res: List[int] = []

        for i in range(n - 1, -1, -1):
            num = nums[i]
            if maxv % num == 0 and maxc == dp[i]:
                maxv = num
                maxc -= 1
                res.append(num)

        return res[::-1]
